In [1]:
import pandas as pd
import numpy as np

lets first read our sample dataset which is in csv file

In [2]:
dataset = pd.read_csv('./data/spam.csv')

In [3]:

dataset.head()

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In [4]:
dataset.isnull().sum()

Category    0
Message     0
dtype: int64

In [5]:
dataset.describe

<bound method NDFrame.describe of      Category                                            Message
0         ham  Go until jurong point, crazy.. Available only ...
1         ham                      Ok lar... Joking wif u oni...
2        spam  Free entry in 2 a wkly comp to win FA Cup fina...
3         ham  U dun say so early hor... U c already then say...
4         ham  Nah I don't think he goes to usf, he lives aro...
...       ...                                                ...
5567     spam  This is the 2nd time we have tried 2 contact u...
5568      ham               Will ü b going to esplanade fr home?
5569      ham  Pity, * was in mood for that. So...any other s...
5570      ham  The guy did some bitching but I acted like i'd...
5571      ham                         Rofl. Its true to its name

[5572 rows x 2 columns]>

This dataset has the following count of ham and spam text.
count of ham    = 4825 
count of 
spam   =   747

In [6]:
dataset['Category'].value_counts()

Category
ham     4825
spam     747
Name: count, dtype: int64

lets replace ham with 0 and spam with 1 

In [7]:
dataset.replace({'Category': {'ham': 0, 'spam': 1}}, inplace=True)
 

In [8]:
#import matplotlib.pyplot as plt
#import seaborn as sns
#plt.figure(figsize=(14,6))
#sns.set_style('darkgrid')
#sns.countplot(x=dataset['Category'].value_counts(), data=dataset)
#plt.title('Number of Spam and Ham')

In [9]:
dataset.describe

<bound method NDFrame.describe of       Category                                            Message
0            0  Go until jurong point, crazy.. Available only ...
1            0                      Ok lar... Joking wif u oni...
2            1  Free entry in 2 a wkly comp to win FA Cup fina...
3            0  U dun say so early hor... U c already then say...
4            0  Nah I don't think he goes to usf, he lives aro...
...        ...                                                ...
5567         1  This is the 2nd time we have tried 2 contact u...
5568         0               Will ü b going to esplanade fr home?
5569         0  Pity, * was in mood for that. So...any other s...
5570         0  The guy did some bitching but I acted like i'd...
5571         0                         Rofl. Its true to its name

[5572 rows x 2 columns]>

In [10]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(dataset['Message'], dataset['Category'], test_size=0.2, random_state=3)

In [11]:
X_train

3075                  Don know. I did't msg him recently.
1787    Do you know why god created gap between your f...
1614                         Thnx dude. u guys out 2nite?
4304                                      Yup i'm free...
3266    44 7732584351, Do you want a New Nokia 3510i c...
                              ...                        
789     5 Free Top Polyphonic Tones call 087018728737,...
968     What do u want when i come back?.a beautiful n...
1667    Guess who spent all last night phasing in and ...
3321    Eh sorry leh... I din c ur msg. Not sad alread...
1688    Free Top ringtone -sub to weekly ringtone-get ...
Name: Message, Length: 4457, dtype: object

In [12]:
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer()
train_vector = cv.fit_transform(X_train)

In [13]:
train_vector

<4457x7694 sparse matrix of type '<class 'numpy.int64'>'
	with 59126 stored elements in Compressed Sparse Row format>

In [14]:
X_train.shape

(4457,)

In [15]:
train_vector.shape

(4457, 7694)

Lets apply MultinomialNB from sklearn.naive_bayes 

In [16]:
from sklearn.naive_bayes import MultinomialNB

classifier = MultinomialNB()
classifier.fit(train_vector, y_train)

MultinomialNB()

In [17]:
from sklearn.metrics import accuracy_score

ytest_pred = classifier.predict(cv.transform(X_test))
print(accuracy_score(y_test, ytest_pred))

ytrain_pred = classifier.predict(cv.transform(X_train))
print(accuracy_score(y_train, ytrain_pred))

0.9901345291479821
0.9928202827013687


Applying Guassian Naive Bayes

In [18]:
from sklearn.naive_bayes import GaussianNB
bayes_classifier = GaussianNB()
bayes_classifier.fit(train_vector.toarray(), y_train)


GaussianNB()

In [19]:
# Predicting
y_pred = bayes_classifier.predict(cv.transform(X_test).toarray())
print(accuracy_score(y_test, ytest_pred))

ytrain_pred = classifier.predict(cv.transform(X_train).toarray())
print(accuracy_score(y_train, ytrain_pred))

0.9901345291479821
0.9928202827013687


In [20]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier(random_state=50)
dt.fit(train_vector.toarray(), y_train)

DecisionTreeClassifier(random_state=50)

In [21]:
from sklearn.tree import DecisionTreeClassifier
DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0,
            min_samples_leaf=1, 
            min_weight_fraction_leaf=0.0, random_state=50,
            splitter='best')

DecisionTreeClassifier(random_state=50)

In [22]:
# Predicting
y_pred_dt = dt.predict(cv.transform(X_test).toarray())

In [23]:
from sklearn.metrics import confusion_matrix
# Evaluating
cm = confusion_matrix(y_test, y_pred_dt)

print(cm)

[[948  12]
 [ 21 134]]


In [24]:
print ("Accuracy : %0.5f \n\n" % accuracy_score(y_test, y_pred_dt))

Accuracy : 0.97040 




In [25]:
from sklearn.metrics import classification_report
print (classification_report(y_test, y_pred_dt) )

              precision    recall  f1-score   support

           0       0.98      0.99      0.98       960
           1       0.92      0.86      0.89       155

    accuracy                           0.97      1115
   macro avg       0.95      0.93      0.94      1115
weighted avg       0.97      0.97      0.97      1115



In [26]:
# import Random Forest classifier

from sklearn.ensemble import RandomForestClassifier

# instantiate the classifier 
rfc = RandomForestClassifier(random_state=0)
# fit the model
rfc.fit(train_vector, y_train)

RandomForestClassifier(random_state=0)

In [27]:
# Predict the Test set results
y_pred = rfc.predict(cv.transform(X_test))

In [28]:
#Check accuracy score 
from sklearn.metrics import accuracy_score
print('Model accuracy score with 10 decision-trees : {0:0.4f}'. format(accuracy_score(y_test, y_pred)))

Model accuracy score with 10 decision-trees : 0.9749


#########################################################################

Comparison and Final Accuracy
DecisionTree  = 97.040 %
Random Forest = 97.49 %
GaussianNB = 99.28 %
MultinomialNB = 99.28 %

#########################################################################

The pickle module implements binary protocols for serializing and de-serializing a Python object structure.

In [29]:
import pickle
# Save model to file
with open('model.pkl', 'wb') as file:
    pickle.dump(classifier, file)


In [30]:
# Load model from file
with open('model.pkl', 'rb') as file:
    loaded_model = pickle.load(file)

In [31]:
ytest_pred = loaded_model.predict(cv.transform(X_test))
print(classification_report(y_test, ytest_pred))

              precision    recall  f1-score   support

           0       0.99      1.00      0.99       960
           1       0.99      0.94      0.96       155

    accuracy                           0.99      1115
   macro avg       0.99      0.97      0.98      1115
weighted avg       0.99      0.99      0.99      1115



In [32]:
# Save the CountVectorizer object to a file
with open('count_vectorizer.pkl', 'wb') as file:
    pickle.dump(cv, file)

In [33]:
# Load the CountVectorizer object from the file
with open('count_vectorizer.pkl', 'rb') as file:
    loaded_cv = pickle.load(file)

In [34]:
ytest_pred = loaded_model.predict(loaded_cv.transform(X_test))
print(classification_report(y_test, ytest_pred))

              precision    recall  f1-score   support

           0       0.99      1.00      0.99       960
           1       0.99      0.94      0.96       155

    accuracy                           0.99      1115
   macro avg       0.99      0.97      0.98      1115
weighted avg       0.99      0.99      0.99      1115



Let's take some examples to test what we did earlier

In [35]:
messages = ["lets catchup at lunch", "Reactivate your account","Congratulations! You have won a free cruise!", "You won 100 euro in a loto ", "Hello darling, what are you up to tonight?", 
           "Bitcoin, anyone?", "You have a new billing statement"]


In [36]:
new_messages = loaded_cv.transform(messages)
predictions = loaded_model.predict(new_messages)
print("here are the required results:")
print(predictions)

here are the required results:
[0 0 1 1 0 0 1]
